crates/jstz_node/src/services/logs/mod.rs

```diff
@@ -1,85 +1,204 @@
+        Address::from_base58(&address).map_err(|e| io::Error::new(InvalidInput, e))?;
 
     Ok(broadcaster.new_client(address).await)
 }
 
+#[derive(Deserialize, Debug)]
+pub struct Pagination {
+    limit: Option<usize>,
+    offset: Option<usize>,
+}
+
+#[get("{address}/persistent/requests")]
+async fn persistent_logs(
+    pagination: Query<Pagination>,
+    pool: Data<SqliteConnectionPool>,
+    path: Path<String>,
+) -> anyhow::Result<HttpResponse, actix_web::Error> {
+    let address = path.into_inner();
+
+    let address =
+        Address::from_base58(&address).map_err(|e| io::Error::new(InvalidInput, e))?;
+
+    let Pagination { limit, offset } = pagination.into_inner();
+    let result = query(
+        &pool,
+        QueryParams::GetLogsByAddress(
+            address,
+            limit.unwrap_or(DEAULT_PAGINATION_LIMIT),
+            offset.unwrap_or(DEAULT_PAGINATION_OFFSET),
+        ),
+    )
+    .await?;
+
+    Ok(HttpResponse::Ok().json(result))
+}
+
+#[get("{address}/persistent/requests/{request_id}")]
+async fn persistent_logs_by_request_id(
+    pool: Data<SqliteConnectionPool>,
+    path: Path<(String, String)>,
+) -> anyhow::Result<HttpResponse, actix_web::Error> {
+    let (address, request_id) = path.into_inner();
+
+    let address =
+        Address::from_base58(&address).map_err(|e| io::Error::new(InvalidInput, e))?;
+
+    let result = query(
+        &pool,
+        QueryParams::GetLogsByAddressAndRequestId(address, request_id),
+    )
+    .await?;
+
+    Ok(HttpResponse::Ok().json(result))
+}
+
 pub struct LogsService;
 
 impl Service for LogsService {
     fn configure(cfg: &mut ServiceConfig) {
-        let scope = Scope::new("/logs").service(stream_logs);
+        let scope = Scope::new("/logs")
+            .service(stream_logs)
+            .service(persistent_logs)
+            .service(persistent_logs_by_request_id);
+
         cfg.service(scope);
     }
 }
 
+// Represents each line in the log file.
+pub enum Line {
+    // Indicates the start and end of a smart function call (request).
+    Request(RequestEvent),
+    // Indicates the js log message from the smart function (e.g. log).
+    Js(LogRecord),
+}
+
 impl LogsService {
-    // Initalise the LogService by spawning a future that reads and broadcasts the file
-    pub fn init(
-        path: String,
+    // Initalise the LogService by spawning a future that reads the file.
+    // The content of the file is broadcasted to clients and flushed to storage.
+    pub async fn init(
+        log_file_path: String,
         cancellation_token: &CancellationToken,
-    ) -> (Arc<Broadcaster>, JoinHandle<Result<()>>) {
+    ) -> anyhow::Result<(
+        Arc<Broadcaster>,
+        SqliteConnectionPool,
+        JoinHandle<io::Result<()>>,
+    )> {
+        // Create a broadcaster for streaming logs.
         let broadcaster = Broadcaster::create();
+        let broadcaster_arc = Arc::clone(&broadcaster);
```

Comment:
Would be good to split `tail_file_handler` into a separate function in which we call `rt::spawn` on to run it -- this would remove the need for `broadcaster_arc` + make any moves more explicit.

Reply:
much cleaner!